### PR TITLE
Move locking from bootstrap.py to rust bootstrap, using fd-lock

### PR DESCRIPTION
**Pull request by joshtriplett**
Helps with https://github.com/rust-lang/rust/issues/94829.

### CODE REVIEW

File: src/bootstrap/bin/main.rs
Comment (jyn514) [inline src/bootstrap/bin/main.rs:25]: The error handling convention in bootstrap is the `t!` macro, which exits with a helpful message on error.
```suggestion
        build_lock = fd_lock::RwLock::new(t!(std::fs::File::create(config.out.join("lock"))));
        _build_lock_guard = match build_lock.try_write() {
            Ok(lock) => lock,
            err => {
                println!("warning: build directory locked, waiting for lock");
                drop(err);
                t!(build_lock.write())
```

File: src/bootstrap/bin/main.rs
Comment (jyn514) [inline src/bootstrap/bin/main.rs:24]: I think you can avoid this drop by using a wildcard:
```suggestion
            Err(_) => {
                println!("warning: build directory locked, waiting for lock");
```

Comment (joshtriplett) [reply src/bootstrap/bin/main.rs:24]: @jyn514 That was the previous thing I tried, but if I do that, Rust still thinks the borrow exists (even though the borrow is in the `Ok` variant, Rust doesn't understand that).

File: src/bootstrap/bin/main.rs
Comment (joshtriplett) [reply src/bootstrap/bin/main.rs:25]: `t` appears to be crate-private to the `bootstrap` library crate; I can move it to `pub`, if you'd like it used in the binary too.

File: src/bootstrap/bin/main.rs
Comment (jyn514) [reply src/bootstrap/bin/main.rs:24]: Ok, fair enough

**Review by jyn514 — COMMENTED** [review]
This is great, thanks! I have a bad feeling CI will fail because of the new dependencies, but r=me otherwise. If it does fail, we should probably have a discussion with Mark / infra team about whether the dependencies are ok.

**Comment by jyn514** [comment]
Oh, last thing - can you see what the compile times for bootstrap before and after this PR? You can clear only the bootstrap artifacts (and not the stage0 toolchain) with `rm -rf build/bootstrap`.

**Comment by joshtriplett** [comment]
On my system, 16.39s before, 16.83s after. I think the new dependencies build almost entirely in parallel with other things.

**Review by jyn514 — APPROVED** [review]

**Comment by Mark-Simulacrum** [comment]
I'm sad we need this many dependencies added, but I suppose they're all fast to compile.

I am a little worried that this is moving the locking into rustbuild, which means that (for example) submodule management is no longer covered by it -- probably amongst several other similar things. Are we confident those behave well when racing two different builds?

File: src/bootstrap/bin/main.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/bin/main.rs:25]: Does the lock guard here get correctly cleared after rustbuild is killed, including with process::exit(...) or ctrl+c?

Both of those are today expected to happen in typical usage, I think.

**Comment by jyn514** [comment]
> submodule management is no longer covered by it -- probably amongst several other similar things. Are we confident those behave well when racing two different builds

Submodule handling is being moved out of python in #97513, happy to wait until that's merged to land this. The only other major things are downloading the bootstrap toolchain and building rustbuild. 

Downloading the toolchain is mostly atomic: we download the files to a temporary directory, then atomically rename them to somewhere in build/cache. Extracting is not atomic but the chance you try and extract the same *destination* file at once seems vanishingly small - IMO the worst scenario in practice for both of these is doing extra work, which seems ~fine given that this is already best effort.

Cargo already has its own locks for building rustbuild, I'm not worried about that.

File: src/bootstrap/bin/main.rs
Comment (joshtriplett) [reply src/bootstrap/bin/main.rs:25]: @Mark-Simulacrum Yes, OS file locking doesn't persist after the file is closed by the process, so if the process exits the lock goes away. This is the major advantage of OS file locking over just using the existence of a lockfile.

File: src/bootstrap/bin/main.rs
Comment (joshtriplett) [reply src/bootstrap/bin/main.rs:19]: `File::create` allows creating the file if it doesn't exist, but will use an existing file if one already exists. So this just ensures the existence of the file.

Comment (Mark-Simulacrum) [reply src/bootstrap/bin/main.rs:19]: Hm, I guess it truncates, which is why I assumed it might be a new file. But sounds like it should be ok.

**Pull request merged**
